Add a File Uploader Widget

dcaminos opened the pull request:
https://github.com/streamlit/streamlit/issues/120

Example:

```
file_png = st.file_uploader("Upload a PNG image", type=([".png"]))

if file_png:
    file_png_bytes = st.file_reader(file_png)
    st.image(file_png_bytes)

```
<img width="1680" alt="Screen Shot 2019-10-28 at 12 04 18 PM (2)" src="https://user-images.githubusercontent.com/22157341/67690262-5b537300-f97b-11e9-850a-a157de496db8.png">

dcaminos commented:
All documentation is missing!!! WIP

tconkling commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,90 @@
+        file = self.fileList.get(widgetId)
+        if file != None:
+            file.buffers[index] = data
+            file.missingChunks = file.missingChunks - 1
```
Also, I'm not sure that `file.missingChunks` is valuable here. What about having it be a class property that is calculated on demand, by counting the number of chunks that are empty? That way we don't have a variable that could become incorrect (for example, if we receive the same buffer twice).

(This is a nit, and is my personal pref, so no need to change if you disagree!)

tconkling commented on proto/streamlit/proto/FileReader.proto:
```diff
@@ -0,0 +1,22 @@
+message FileReader {
+  string id = 1;
+  string file = 2;
+}
```
Add newline at end of file

tconkling commented on proto/streamlit/proto/FileUploader.proto:
```diff
@@ -0,0 +1,24 @@
+  string label = 2;
+  repeated string type = 3;
+  int32 max_upload_size = 4;
+}
```
Add newline at end of file

tconkling commented on proto/streamlit/proto/FileReader.proto:
```diff
@@ -0,0 +1,22 @@
+
+syntax = "proto3";
+
+message FileReader {
```
Add comments describing what "id" and "file" are. (Looks like "file" is the _name_ of the file, and not the file itself? If so, then "filename" might be better?)

dcaminos replied:
Is not the file name, is the el fileID devuelto por el file_uploader. Pero puedo cambiarlo :)

tconkling commented on proto/streamlit/proto/FileUploadStatus.proto:
```diff
@@ -0,0 +1,34 @@
+
+syntax = "proto3";
+
+// A python empty.
```
?

tconkling commented on lib/streamlit/ReportSession.py:
```diff
@@ -107,6 +109,8 @@ def __init__(self, ioloop, script_path, command_line):
 
         self._scriptrunner = None
 
+        self.fileManager = FileManager()
```
should be `self._file_manager` (snake-case, and also a leading underscore since it's a private variable)

tconkling commented on frontend/src/lib/WidgetStateManager.ts:
```diff
@@ -193,6 +194,32 @@ export class WidgetStateManager {
+  ): void {
+    const limit = 10 * 1e6 //10MB
+
+    const newFileMessage: NewFile = new NewFile()
```
nit: `const newFileMessage = new NewFile()` (no need to specify the type twice, since it can be inferred from the constructor)

tconkling commented on frontend/src/lib/WidgetStateManager.ts:
```diff
@@ -193,6 +194,32 @@ export class WidgetStateManager {
+    this.sendBackMsg({ newFile: newFileMessage })
+
+    for (let i = 0; i < newFileMessage.chunks; i++) {
+      const message: FileChunk = new FileChunk()
```
nit: `const message = new FileChunk()` (actually I think our ESlint should be catching/fixing these? maybe it's not running properly?)

dcaminos commented on frontend/src/App.tsx:
```diff
@@ -448,6 +464,20 @@ class App extends PureComponent<Props, State> {
     }
   }
 
+  updateFileUploaderWidget = (
```
Was an old implementation, I forgot delete it :)

tvst commented on examples/file_uploader.py:
```diff
@@ -0,0 +1,48 @@
+from __future__ import print_function, division, unicode_literals, absolute_import
+from streamlit.compatibility import setup_2_3_shims
+
+setup_2_3_shims(globals())
```
Remove lines 18-22, unless you really need them.

We're trying to stop loading this compatibility libraries because they're often hard to debug. And most times they can be dropped without any problems.

tvst commented:
Found a bug: the file uploader gets into a "loading forever" state when you pass the wrong kind of file.

Steps to repro:
1) Run examples/file_uploader.py
2) In the CSV file uploader, pass in an MP3 file
**Expected:** the file uploads, but the script throws an exception that appears in the app when it tries to read the MP3 as a CSV
**Actual:** the file uploader never finishes uploading the file

dcaminos commented:
@tconkling I fixed all your comments and added a test for `FileManager` (the most important one). `ReportSession` and `Cypress` tests are missing yet.

@tvst I fixed that bug and added a method to cancel the upload.

tvst commented on proto/streamlit/proto/BackMsg.proto:
```diff
@@ -49,5 +67,14 @@ message BackMsg {
+    NewFile new_file = 11;
+
+    // Upload a file chunk to the server
+    FileChunk file_chunk = 12;
```
Rename to `upload_file_chunk`

tvst commented on proto/streamlit/proto/BackMsg.proto:
```diff
@@ -49,5 +67,14 @@ message BackMsg {
     bool close_connection = 10;
+
+    // Start to upload a new file to the server
+    NewFile new_file = 11;
```
Rename to `upload_file`

tvst commented on proto/streamlit/proto/Element.proto:
```diff
@@ -76,7 +80,7 @@ message Element {
     VegaLiteChart vega_lite_chart = 10;
     Video video = 14;
-    NumberInput number_input = 29;
+    NumberInput number_input = 31;
```
Don't renumber fields. This should stay as `29`, and the new ones should be `30` and `31`.

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1841,6 +1843,75 @@ def slider(
+        --------
+        >>> file = st.file_uploader("Upload a image", type=([".png"]))
+        >>> if file:
+        >>>     file_bytes = st.file_reader(file)
```
Should be `st.file_uploader`

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1841,6 +1843,75 @@ def slider(
+        return ui_value if ui_value is not None else ""
+
+
+    def file_reader(self, file_id):
```
I don't understand: what is this? Is this another widget type? Or just some leftover code?

tvst commented on lib/streamlit/__init__.py:
```diff
@@ -157,6 +157,8 @@ def _reset(main_dg, sidebar_dg):
 empty = _with_dg(_DeltaGenerator.empty)  # noqa: E221
 error = _with_dg(_DeltaGenerator.error)  # noqa: E221
 exception = _with_dg(_DeltaGenerator.exception)  # noqa: E221
+file_reader = _with_dg(_DeltaGenerator.file_reader)  # noqa: E221
```
Same here. Is this leftover code?

tvst commented on proto/streamlit/proto/Element.proto:
```diff
@@ -62,6 +64,8 @@ message Element {
     DocString doc_string = 7;
     Empty empty = 2;
     Exception exception = 8;
+    FileReader file_reader = 30;
```
Same here: is this leftover code?

tvst commented on proto/streamlit/proto/FileReader.proto:
```diff
@@ -0,0 +1,28 @@
+
+syntax = "proto3";
+
+// file_reader api
```
This comment doesn't add any new information. Remove.

tvst commented on proto/streamlit/proto/FileReader.proto:
```diff
@@ -0,0 +1,28 @@
+// file_reader api
+// When upload a file with the file_uploader widget, you get a fileId.
+// With this endpoint you get the bytes of that file 
+message FileReader {
```
Same here: is this leftover code?

tvst commented on proto/streamlit/proto/ForwardMsg.proto:
```diff
@@ -72,6 +73,9 @@ message ForwardMsg {
     // in its cache, it can retrieve it from the server.
     string ref_hash = 11;
+
+
```
Too many newlines. Should be only 1 blank line.

tvst commented on lib/tests/streamlit/FileManager_test.py:
```diff
@@ -0,0 +1,60 @@
+            f.close()
+            self.assertEqual(file_bytes, data)
+
+        
```
Nit: should be only 1 blank line.

tvst commented on lib/tests/streamlit/FileManager_test.py:
```diff
@@ -0,0 +1,60 @@
+class FileManagerTest(unittest.TestCase):
+    def test_msg_hash(self):
+        
+        
```
Remove all these blank lines.

dcaminos commented on frontend/src/App.tsx:
```diff
@@ -259,6 +261,20 @@ class App extends PureComponent<Props, State> {
+            fileUploadStatus.state === FileUploadStatus.FileUploadState.DELETED
+          ) {
+            // FINISHED TO UPLOAD
+            this.widgetMgr.setStringValue(
```
I changed this implementation. Now, I rerun the delta with the new file value to send it to the UI

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1876,6 +1881,39 @@ def slider(
+        _set_widget_id("file_uploader", element, user_key=key)
+        progress, data = self._file_manager.get_data(element.file_uploader.id)
+        element.file_uploader.progress = progress
+        return data if data != None else NoValue
```
Flip the order of this `if`, otherwise the `else` is a double-negative, which is hard to read.

Also, for `None` use `is not None` instead of `!= None`

```py
return NoValue if data is None else data
```

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1876,6 +1881,39 @@ def slider(
+        _set_widget_id("file_uploader", element, user_key=key)
+        progress, data = self._file_manager.get_data(element.file_uploader.id)
+        element.file_uploader.progress = progress
+        return data if data != None else NoValue
```
Also, this `NoValue` thing doesn't seem to be working for me.

When I run this:

```py
import streamlit as st

the_file = st.file_uploader("Upload a file")
'Result:', the_file
```

I get:

![image](https://user-images.githubusercontent.com/690814/68443651-75fec680-0189-11ea-9294-1565e67ab45a.png)

But I expected to get `Result: None`

dcaminos replied:
This is weird. It's working for me:
<img width="718" alt="Screen Shot 2019-11-08 at 12 46 29 PM" src="https://user-images.githubusercontent.com/22157341/68511759-664cb400-0255-11ea-9686-3672e79bcef9.png">

dcaminos commented on proto/streamlit/proto/Element.proto:
```diff
@@ -49,35 +50,37 @@ message Element {
     NumberInput number_input = 29;
-
-    // Next ID: 29
+    FileUploader file_uploader = 30;
```
Sorry, I was checking all ids, I forgot back it to normal.

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1876,6 +1881,39 @@ def slider(
+
+        Examples
+        --------
+        >>> file = st.file_uploader("Upload a image", type=([".png"]))
```
Should be `type=[".png"]`

But actually, can we change `type` so:
* It works if you pass a `str`. (That is, it should accept both `str` and `iterable`)
* You don't need to pass in the `.` in `.png`

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1876,6 +1881,39 @@ def slider(
+            A short label explaining to the user what this file uploader is for.
+        type : string[] or None
+            Array of allowed extensions. (['.png', '.jpg'])
+            Default value: ['*']
```
This means `type` is optional, right? (Which I think is the right approach)

However, if I don't pass a type, and then I try to upload a file, I get an error.

Steps to repro:
1) `st.file_uploader("Upload a file")
2) run the script
3) upload an SVG file, for example

**Expected:** it should just work
**Result:** this error shows up `image/svg+xml files are not allowed`


When  like "image/svg+xml files are not allowed"

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,90 @@
+
+    def porcess_chunk(self, widgetId, index, data):
+        file = self.fileList.get(widgetId)
+        if file != None:
```
(Just making sure you saw this)

dcaminos replied:
Yes, each thing I fix, I resolve the conversation. I will do this one right now.

dcaminos replied:
When you cancel an upload, maybe a chunk was sended, and the file was deleted. So this should be a normal behavior.

tvst replied:
Then flip the `if` around and add a comment explaining this, please.

Also, I think the FileManager shouldn't know how to add chunks to a File. Instead it should call a method `file.process_chunk(...)` which knows what to do.

So when you add all this together, this method would become simply this:

```py
if widget_id not in self._file_list:
  # Handle possible race condition when you cancel an upload
  # and an old file chunk is received.
  return

return self._file_list[widget_id].process_chunk(index, data)
```

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,90 @@
+    def porcess_chunk(self, widgetId, index, data):
+        file = self.fileList.get(widgetId)
+        if file != None:
+            file.buffers[index] = data
```
(Just making sure you saw this)

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1876,6 +1881,39 @@ def slider(
+    @_with_element
+    def file_uploader(self, element, label, type=['*'], key=None):
+
+        """Display a file uploader widget.
```
Add some text talking about the size limit:

```
By default, uploaded files are limited to 50MB but you can configure that using the `foo.bar` config option.
```

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1876,6 +1881,39 @@ def slider(
+            A short label explaining to the user what this file uploader is for.
+        type : string[] or None
+            Array of allowed extensions. (['.png', '.jpg'])
+            Default value: ['*']
```
What does `type=None` mean? Is that the same as `['*']`?

I think we shouldn't have `['*']` and the default should be `None` — which would mean that _all_ file types are allowed.

treuille replied:
Also, why are there periods in the name. Adds nothing, right?

treuille commented:
I'm so excited to see this feature in Streamlit!!

sssdjj commented:
> #120
> 
> Example:
> 
> ```
> file_png = st.file_uploader("Upload a PNG image", type=([".png"]))
> 
> if file_png:
>     file_png_bytes = st.file_reader(file_png)
>     st.image(file_png_bytes)
> ```
> 
> <img alt="Screen Shot 2019-10-28 at 12 04 18 PM (2)" width="1680" src="https://user-images.githubusercontent.com/22157341/67690262-5b537300-f97b-11e9-850a-a157de496db8.png">

the 0.5.2 have upload?

dcaminos commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -264,7 +266,8 @@ class DeltaGenerator(object):
       The full path of this DeltaGenerator, consisting of the IDs of
       all ancestors. The 0th item is the topmost ancestor.
-
+    file_manager: File_manager
```
I changed by `get_report_ctx` solution, we can talk later which way is better but we all agree should work in the same way for both (FileManager and WidgetState)

treuille commented:
Not yet, @sssdjj ! Hopefully the next release!

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1936,18 +1933,30 @@ def file_uploader(self, element, label, type=['*'], key=None):
+                else:
+                    element.file_uploader.type[:] = ['.' + ''.join(type)] # example: type="csv"
+            else:
+                element.file_uploader.type[:] = map(lambda x: x if x[0] == '.' else '.' + x , type) # example: type=["png", ".jpeg"]
```
I don't understand why this logic is needed. The file type passed by the user should **never** have a period.

So you could replace the whole logic above with just:

```py
if _type(type) in string_types:
  type = [type]

element.file_uploader.type[:] = [
   "." + t for t in type
]
```

Also: this is a bit of a nit-pick, but I think the best place to add a period to the types is actually in JS rather than in Python. Otherwise you're exposing a JS implementation detail to Python for no good reason. So I'd just replace this whole thing with:

```py
if _type(type) in string_types:
  type = [type]

element.file_uploader.type[:] = type
```

If it turns out to be annoying to do this in JS, though, feel free to use the first code snippet above instead.

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1936,18 +1933,30 @@ def file_uploader(self, element, label, type=['*'], key=None):
+            progress, data = ctx.file_manager.get_data(element.file_uploader.id)
+            element.file_uploader.progress = progress
+
+        return NoValue if data is None else data
```
There's a bug here. When `ctx` is `None`, `data` will not be declared and you'll get an exception.

tvst commented on lib/streamlit/ScriptRunner.py:
```diff
@@ -48,7 +48,7 @@ class ScriptRunnerEvent(Enum):
 
 class ScriptRunner(object):
-    def __init__(self, report, main_dg, sidebar_dg, widget_states, request_queue):
+    def __init__(self, report, main_dg, sidebar_dg, widget_states, request_queue, file_manager=None):
```
Add `file_manager` to the pydoc below.

tvst commented on lib/streamlit/ReportSession.py:
```diff
@@ -431,6 +433,29 @@ def handle_rerun_script_request(
 
         self.request_rerun(widget_state)
 
+    def handle_upload_file(self, upload_file=None):
```
Why doss `upload_file` have a default value? The only place that calls `upload_file()` actually passes this argument, so the default is no needed (and confusing)

dcaminos replied:
I did it because all others methods called in server.py use default params. But You are right, I will remove it.

tvst commented on lib/streamlit/ReportSession.py:
```diff
@@ -431,6 +433,29 @@ def handle_rerun_script_request(
+
+        self.handle_rerun_script_request(widget_state=self._widget_states)
+
+    def handle_upload_file_chunk(self, upload_file_chunk=None):
```
Same here. `upload_file_chunk` should have no default value.

tvst commented on lib/streamlit/ReportSession.py:
```diff
@@ -431,6 +433,29 @@ def handle_rerun_script_request(
+        if progress==1:
+            self.handle_rerun_script_request(widget_state=self._widget_states)
+
+    def handle_delete_uploaded_file(self, delete_uploaded_file=None):
```
...and here too.

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+    def delete_file(self, widget_id):
+        del self._file_list[widget_id]
+
+    def locate_new_file(self, widget_id, name, size, last_modified, chunks):
```
The name of this function is really odd. From what I understand, this should be called `create_or_clear_file`

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+            if len(file.buffers) == file.total_chunks:
+                file.data = bytearray()
+                index = 0
+                while file.buffers.get(index) != None:
```
Why not just use `for index in range(len(file.buffers))` ?

dcaminos replied:
Because we don't know what indexes have the buffer dic. maybe the chunk 3 come first.

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+
+    def get_data(self, widget_id):
+        file = self._file_list.get(widget_id)
+        if file != None:
```
Nit: use the `in` operator for clarity

```py
if widget_id not in self._file_list:
  return 0, None

file = self._file_list[widget_id]

if file.data:
  return 1, file.data

progress = float(len(file.buffers)) / file.total_chunks
return progress, None
```

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+    def locate_new_file(self, widget_id, name, size, last_modified, chunks):
+        file = self._file_list.get(widget_id)
+        if file != None:
+            self.delete_file(widget_id)
```
The 3 lines above would be clearer if you replaced with this:

```py
if widget_id in self._file_list:
  self.delete_file(widget_id)
```

...because this actually reads like what you want to happen: "if the widget ID exists in the file list, delete the file with that widget ID".

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+        file = self._file_list.get(widget_id)
+        if file != None:
+            if file.buffers.get(index) is not None:
+                LOGGER.error("File chunk was already processed")
```
Use the `in` operator for clarity: `if index in file.buffers:`

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+        )
+        self._file_list[widget_id] = file
+
+    def process_chunk(self, widget_id, index, data):
```
Add a pydoc since the return value of this function is non-obvious:

```py
"""Process an incoming file chunk and return percent done."""
```

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+                return 1
+
+            if len(file.buffers) > 0:
+                return float(len(file.buffers))/float(file.total_chunks)
```
You only need to convert one of these to float, not both.

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+                    del file.buffers[index]
+                    index += 1
+
+                file.buffers = None
```
Shouldn't this be `{}`?

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+        file = self._file_list.get(widget_id)
+        if file != None:
+            if file.buffers.get(index) is not None:
+                LOGGER.error("File chunk was already processed")
```
Sounds like we should throw an exception here. This should _never ever_ happen, right?

dcaminos replied:
No really, it's hard to reproduce, never should happen. We talked about this here: https://github.com/streamlit/streamlit/pull/488#discussion_r338248994

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+                LOGGER.error("File chunk was already processed")
+
+            file.buffers[index] = data
+            if len(file.buffers) == file.total_chunks:
```
Move everything inside this `if` to a `_coalesce_chunks()` method for readability:

```py
if len(file.buffers) == this.total_chunks:
  self._coalesce_chunks()
```

(the code snippet above assumes this whole logic lives inside `File`, as described in [this](https://github.com/streamlit/streamlit/pull/488/files#r338248577) comment thread.

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+
+        return 0
+
+    def get_data(self, widget_id):
```
Add a pydoc since the return value of this method is notrivial:

```py
"""Get a tuple with file progress and data bytes (or None)."""
```

tvst commented on lib/streamlit/fileManager.py:
```diff
@@ -0,0 +1,88 @@
+    def get_data(self, widget_id):
+        file = self._file_list.get(widget_id)
+        if file != None:
+            if file.data is not None:
```
`file.data` is either None or some bytes. So you can just replace this with `if file.data` for clarity

tvst commented:
🐞 Found a bug. 

1. Run the code below:
   ```
   import streamlit as st

   the_file = st.file_uploader("Upload a file", type=[".pdf"])

   if the_file is None:
       'Result:', None
   else:
       'Result:', len(the_file)
   ```

2. Upload a PDF file
3. Change the `type` in the code:
   ```
   the_file = st.file_uploader("Upload a file", type=[".zip"])
   ```
4. In the app, press the "rerun" button in the top right.

### Expected: 
![image](https://user-images.githubusercontent.com/690814/68836038-47dc2380-066e-11ea-8691-5d37a1e6ee31.png)

### Actual:
![image](https://user-images.githubusercontent.com/690814/68836102-69d5a600-066e-11ea-8b27-0d5eaebf3ec6.png)

tvst commented on frontend/src/components/widgets/FileUploader/FileUploader.tsx:
```diff
@@ -0,0 +1,163 @@
+    const accept: string[] = element.get("type").toArray()
+    const label: string = element.get("label")
+    return (
+      <div className="file-uploader">
```
The `className` should be `Widget stFileUploader`.

Without `Widget`, the `<label>` doesn't look right.

And all Streamlit elements need to be wrapped in a DOM element with class `stElementName` (see our [style guide](https://github.com/streamlit/streamlit/wiki/Style-Guide#css--scss))

dcaminos commented:
> 🐞 Found a bug.
> 
> 1. Run the code below:
>    ```
>    import streamlit as st
>    
>    the_file = st.file_uploader("Upload a file", type=[".pdf"])
>    
>    if the_file is None:
>        'Result:', None
>    else:
>        'Result:', len(the_file)
>    ```
> 2. Upload a PDF file
> 3. Change the `type` in the code:
>    ```
>    the_file = st.file_uploader("Upload a file", type=[".zip"])
>    ```
> 4. In the app, press the "rerun" button in the top right.
> 
> ### Expected:
> ![image](https://user-images.githubusercontent.com/690814/68836038-47dc2380-066e-11ea-8691-5d37a1e6ee31.png)
> 
> ### Actual:
> ![image](https://user-images.githubusercontent.com/690814/68836102-69d5a600-066e-11ea-8b27-0d5eaebf3ec6.png)

With the new changes, I was not able to reproduce this.

ajinkya933 commented:
@dcaminos how do i reproduce your results 

> #120
> 
> Example:
> 
> ```
> file_png = st.file_uploader("Upload a PNG image", type=([".png"]))
> 
> if file_png:
>     file_png_bytes = st.file_reader(file_png)
>     st.image(file_png_bytes)
> ```
> 
> <img alt="Screen Shot 2019-10-28 at 12 04 18 PM (2)" width="1680" src="https://user-images.githubusercontent.com/22157341/67690262-5b537300-f97b-11e9-850a-a157de496db8.png">

Heres what I tried.
I forked your git repo and ran streamlit from there but I constantly get an error of ```st.file_uploader not found```. I am aware that this feature is not yet released but if that is so then how were you able to implement it ?

isConic commented:
👀

tvst commented on proto/streamlit/proto/Element.proto:
```diff
@@ -66,11 +67,13 @@ message Element {
     Json json = 31;
     Markdown markdown = 29;
     MultiSelect multiselect = 28;
+    NumberInput number_input = 33;
```
Please leave number_input as 32. (We should never change proto numbers)

tvst commented on proto/streamlit/proto/FileUploader.proto:
```diff
@@ -0,0 +1,35 @@
+  // Text to be displayed before the widget
+  string label = 2;
+
+  // Supported types: ["png","jpg","img"]
```
I think you mean:

```
  // Supported types. For example: ["png","jpg","img"]
```

tvst commented on proto/streamlit/proto/FileUploader.proto:
```diff
@@ -0,0 +1,35 @@
+  // Max file size allowed by server config
+  int32 max_upload_size_mb = 4;
+
+  // The current progress of upload
```
Add: `Should be from 0 to 1`

tvst commented:
@ajinkya933 : are you sure you're in the `issue120/file-uploader` branch?

ajinkya933 commented:
> @ajinkya933 : are you sure you're in the `issue120/file-uploader` branch?

Yes i am in the `issue120/file-uploader` branch
I am using following file location for demo.py:

```/home/ajinkya/Music/streamlit-issue120-file-uploader/demo.py```

demo.py:

import streamlit as st

file_png = st.file_uploader("Upload a PNG image", type=([".png"]))

if file_png:
    file_png_bytes = st.file_reader(file_png)
    st.image(file_png_bytes)

tconkling commented:
@ajinkya933 Have you installed the `file-uploader` branch into your Python environment? Simply running `demo.py` won't work; assuming you had previously run `pip install streamlit`, then when you `import streamlit as st` in a source file, you're getting the installed release version.

You should probably do something like:
```
$ pip uninstall streamlit
$ pip install -e <path/to/file-uploader/branch/checkout>
```

(The pip documentation contains more details about local project installs: https://pip.pypa.io/en/stable/reference/pip_install/#local-project-installs)

sssdjj commented:
> Not yet, @sssdjj ! Hopefully the next release!
Now, why not update this feature

ajinkya933 commented:
> @ajinkya933 Have you installed the `file-uploader` branch into your Python environment? Simply running `demo.py` won't work; assuming you had previously run `pip install streamlit`, then when you `import streamlit as st` in a source file, you're getting the installed release version of streamlit.
> 
> You should probably do something like:
> 
> ```
> $ pip uninstall streamlit
> $ pip install -e <path/to/file-uploader/branch/checkout>
> ```
> 
> The pip documentation contains more details about local project installs: https://pip.pypa.io/en/stable/reference/pip_install/#local-project-installs

I have followed your steps. And have noticed that I have to install streamlit from develop branch. However there are series of errors I get when I install streamlit develop branch errors:(https://github.com/streamlit/streamlit/issues/728)

I hope someone will help me rectify them

dcaminos commented:
@ajinkya933 that source code is old, you should now use `file_reader` any more (that was my first approach).

Try to run this:

```
import streamlit as st

the_file = st.file_uploader("Upload a file", type=["pdf"])

if the_file is None:
    'Result:', None
else:
    'Result:', len(the_file)
```

ajinkya933 commented:
> @ajinkya933 that source code is old, you should not use `file_reader` any more (that was my first approach).
> 
> Try to run this:
> 
> ```
> import streamlit as st
> 
> the_file = st.file_uploader("Upload a file", type=["pdf"])
> 
> if the_file is None:
>     'Result:', None
> else:
>     'Result:', len(the_file)
> ```

Have you installed streamlit from here: https://github.com/streamlit/streamlit/wiki/Contributing#Ubuntu. 

I am having trouble installing develop version of streamlit as pointed out by @tconkling initially I run

```
ajinkya@ajinkya-H310M-S2:~/Music/streamlit/lib$ pip3 install -e . --user
Obtaining file:///home/ajinkya/Music/streamlit/lib
```

for which I get a message ```Successfully installed streamlit```

But when I do  run ```$ streamlit --version``` i get 
```
$ streamlit --version
Traceback (most recent call last):
  File "/home/ajinkya/.local/bin/streamlit", line 11, in <module>
    load_entry_point('streamlit', 'console_scripts', 'streamlit')()
  File "/home/ajinkya/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 489, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/home/ajinkya/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2852, in load_entry_point
    return ep.load()
  File "/home/ajinkya/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2443, in load
    return self.resolve()
  File "/home/ajinkya/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2449, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/ajinkya/Music/streamlit/lib/streamlit/__init__.py", line 96, in <module>
    from streamlit.DeltaGenerator import DeltaGenerator as _DeltaGenerator
  File "/home/ajinkya/Music/streamlit/lib/streamlit/DeltaGenerator.py", line 35, in <module>
    from streamlit.proto import Balloons_pb2
ImportError: cannot import name 'Balloons_pb2'

```
After getting this error  I followed instructions here (https://github.com/streamlit/streamlit/wiki/Contributing#Ubuntu) and heres my output error
here is the detailed error log (https://github.com/streamlit/streamlit/issues/728)

tvst commented on frontend/src/components/widgets/FileUploader/FileUploader.scss:
```diff
@@ -0,0 +1,68 @@
+  outline: 0;
+}
+
+.stFileUploaderError {
```
We only add `st` to the top-level classname in each component.

Also, wrap the whole CSS in `.stFileUploader {` and then you can just rename this to `.uploadError` or something. For example:

```scss
.stFileUploader {
  ...
  .uploadError {
    ...
  }
  ...
}
```
...same with everything else in this file.

That's what we've been doing in the CSS for other elements.

tvst commented on frontend/src/components/widgets/FileUploader/FileUploader.scss:
```diff
@@ -0,0 +1,68 @@
+  padding-top: 1rem;
+  outline: 0px;
+  -webkit-box-orient: vertical;
+  -webkit-box-direction: normal;
```
Why are all these `-webkit` properties needed? It's best to avoid browser-specific code.

tvst commented on frontend/src/lib/WidgetStateManager.ts:
```diff
@@ -193,6 +196,45 @@ export class WidgetStateManager {
+
+    // The default file size allowed by server config is 50MB.
+    // The best message size to don't freeze the server is 10MB
+    const chunkByteLimit = 10 * 1e6 //10MB
```
What if the user sets the limit to 5MB? Does this break anything?

dcaminos replied:
No, we have a validation to check if the widget accept the file first

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1903,6 +1907,49 @@ def slider(
+        Examples
+        --------
+        >>> file = st.file_uploader("Upload a image", type=["png"])
+        >>> if file != None:
```
`is not None`

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1903,6 +1907,49 @@ def slider(
+
+        Examples
+        --------
+        >>> file = st.file_uploader("Upload a image", type=["png"])
```
`type="png"`

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1903,6 +1907,49 @@ def slider(
+        ----------
+        label : str or None
+            A short label explaining to the user what this file uploader is for.
+        type : list of string or None
```
```
type : str or list of str or None
```

tvst commented on lib/streamlit/folder_black_list.py:
```diff
@@ -32,6 +33,14 @@
 
 
+# Add the Streamlit lib folder when in dev mode, since otherwise we end up with
+# weird situations where the ID of a class in one run is the same as in another
```
Ooops, this should be "is not the same" (I forgot the "not")

universewill commented:
> > @ajinkya933 Have you installed the `file-uploader` branch into your Python environment? Simply running `demo.py` won't work; assuming you had previously run `pip install streamlit`, then when you `import streamlit as st` in a source file, you're getting the installed release version of streamlit.
> > You should probably do something like:
> > ```
> > $ pip uninstall streamlit
> > $ pip install -e <path/to/file-uploader/branch/checkout>
> > ```
> > 
> > 
> > The pip documentation contains more details about local project installs: https://pip.pypa.io/en/stable/reference/pip_install/#local-project-installs
> 
> @tconkling I have followed your steps. And have noticed that I have to install streamlit from `issue120/file-uploader` branch. However there are series of errors I get when I install streamlit develop branch errors:(#728)
> 
> I hope someone will help me rectify them

i try to run ' pip install -e path/to/code' command but it tells me no setup.py found.
How to install?

universewill commented:
> > > @ajinkya933 Have you installed the `file-uploader` branch into your Python environment? Simply running `demo.py` won't work; assuming you had previously run `pip install streamlit`, then when you `import streamlit as st` in a source file, you're getting the installed release version of streamlit.
> > > You should probably do something like:
> > > ```
> > > $ pip uninstall streamlit
> > > $ pip install -e <path/to/file-uploader/branch/checkout>
> > > ```
> > > 
> > > 
> > > The pip documentation contains more details about local project installs: https://pip.pypa.io/en/stable/reference/pip_install/#local-project-installs
> > 
> > 
> > @tconkling I have followed your steps. And have noticed that I have to install streamlit from `issue120/file-uploader` branch. However there are series of errors I get when I install streamlit develop branch errors:(#728)
> > I hope someone will help me rectify them
> 
> i try to run ' pip install -e path/to/code' command but it tells me no setup.py found.
> How to install?

solved by 'cd streamlit/lib'  first

universewill commented:
> > @ajinkya933 that source code is old, you should not use `file_reader` any more (that was my first approach).
> > Try to run this:
> > ```
> > import streamlit as st
> > 
> > the_file = st.file_uploader("Upload a file", type=["pdf"])
> > 
> > if the_file is None:
> >     'Result:', None
> > else:
> >     'Result:', len(the_file)
> > ```
> 
> Have you installed streamlit from here: https://github.com/streamlit/streamlit/wiki/Contributing#Ubuntu.
> 
> I am having trouble installing develop version of streamlit as pointed out by @tconkling initially I run
> 
> ```
> ajinkya@ajinkya-H310M-S2:~/Music/streamlit/lib$ pip3 install -e . --user
> Obtaining file:///home/ajinkya/Music/streamlit/lib
> ```
> 
> for which I get a message `Successfully installed streamlit`
> 
> But when I do run `$ streamlit --version` i get
> 
> ```
> $ streamlit --version
> Traceback (most recent call last):
>   File "/home/ajinkya/.local/bin/streamlit", line 11, in <module>
>     load_entry_point('streamlit', 'console_scripts', 'streamlit')()
>   File "/home/ajinkya/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 489, in load_entry_point
>     return get_distribution(dist).load_entry_point(group, name)
>   File "/home/ajinkya/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2852, in load_entry_point
>     return ep.load()
>   File "/home/ajinkya/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2443, in load
>     return self.resolve()
>   File "/home/ajinkya/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2449, in resolve
>     module = __import__(self.module_name, fromlist=['__name__'], level=0)
>   File "/home/ajinkya/Music/streamlit/lib/streamlit/__init__.py", line 96, in <module>
>     from streamlit.DeltaGenerator import DeltaGenerator as _DeltaGenerator
>   File "/home/ajinkya/Music/streamlit/lib/streamlit/DeltaGenerator.py", line 35, in <module>
>     from streamlit.proto import Balloons_pb2
> ImportError: cannot import name 'Balloons_pb2'
> ```
> 
> After getting this error I followed instructions here (https://github.com/streamlit/streamlit/wiki/Contributing#Ubuntu) and heres my output error
> here is the detailed error log (#728)

get same error

Stamenov commented:
Hey,
I am looking forward to this feature as a very nice addition to the streamlit widgets. 
What is there left to be done? Perhaps I can help.
Best.

ajinkya933 commented:
> Hey,
> I am looking forward to this feature as a very nice addition to the streamlit widgets.
> What is there left to be done? Perhaps I can help.
> Best.

Perhaps you can help me in resolving this issue https://github.com/streamlit/streamlit/issues/728  File uploader feature is available in develop branch but I am having a tough time installing it

jrhone commented on lib/streamlit/caching.py:
```diff
@@ -53,6 +54,18 @@
 
+# based on: https://stackoverflow.com/questions/43506378/how-to-get-source-code-of-function-that-is-wrapped-by-a-decorator
+# In Python 2, the @functools.wraps() decorator does not set the convenience __wrapped__
+if sys.version_info[0:2] >= (3, 4):  # Python v3.4+?
```
Looks like it might be 3.2

> New in version 3.2: Automatic addition of the __wrapped__ attribute.

> To allow access to the original function for introspection and other purposes (e.g. bypassing a caching decorator such as lru_cache()), this function automatically adds a __wrapped__ attribute to the wrapper that refers to the original function.

> Changed in version 3.4: The __wrapped__ attribute now always refers to the wrapped function, even if that function defined a __wrapped__ attribute. (see issue 17482)

jrhone replied:
Oh I just read the comment on stackoverflow, so I see why you chose 3.4

jrhone commented on lib/streamlit/caching.py:
```diff
@@ -53,6 +54,18 @@
+if sys.version_info[0:2] >= (3, 4):  # Python v3.4+?
+    wraps = functools.wraps  # built-in has __wrapped__ attribute
+else:
+    def wraps(wrapped, assigned=functools.WRAPPER_ASSIGNMENTS,
```
Should we add some tests for python 2?

dcaminos replied:
Well, if this doesn't work, a caching test will fail.

dancebean commented:
Is there a way to install the latest version from this branch of repo directly, since this PR has been merged? Since it has not been included in the latest released version, pip install won't get the most updated version

vanpelt commented:
@dancebean `pip install --upgrade git+git://github.com/streamlit/streamlit.git@master#egg=streamlit`

RichardOberdieck commented:
> 
> 
> @dancebean `pip install --upgrade git+git://github.com/streamlit/streamlit.git@master#egg=streamlit`

This breaks when using conda. Is there any indication when this will be released?

universewill commented:
> @dancebean `pip install --upgrade git+git://github.com/streamlit/streamlit.git@master#egg=streamlit`

unable to install this way

treuille commented:
@dancebean, @vanpelt (👋), and @RichardOberdieck : Please note that [`file_uploader` has been released in 0.52](https://streamlit.io/docs/api.html?highlight=file#streamlit.file_uploader) in experimental form.

You can find instructions to build Streamlit from source [here](https://github.com/streamlit/streamlit/wiki/Contributing#how-to-build-streamlit).

Happy New Year! 🎆

universewill commented:
> @dancebean, @vanpelt (👋), and @RichardOberdieck : Please note that [`file_uploader` has been released in 0.52](https://streamlit.io/docs/api.html?highlight=file#streamlit.file_uploader) in experimental form.
> 
> Also, you can find instructions to build Streamlit from source [here](https://github.com/streamlit/streamlit/wiki/Contributing#how-to-build-streamlit).
> 
> Happy New Year! 🎆

Great work! Besides, how to upload an image and write to local as an image file?

universewill commented:
> > @dancebean, @vanpelt (👋), and @RichardOberdieck : Please note that [`file_uploader` has been released in 0.52](https://streamlit.io/docs/api.html?highlight=file#streamlit.file_uploader) in experimental form.
> > Also, you can find instructions to build Streamlit from source [here](https://github.com/streamlit/streamlit/wiki/Contributing#how-to-build-streamlit).
> > Happy New Year! 🎆
> 
> Great work! Besides, how to upload an image and write to local as an image file?

This can do:
'''
    uploaded_file = st.file_uploader("Choose a image file", type=["png", 'jpg', 'jpeg', 'tiff'])
    if uploaded_file is not None:
        im = Image.open(uploaded_file)
        im.save('test.jpg', quality=95)
'''

treuille commented:
Great, @universewill . Please also know that `file_uploader` has an experimental API and may change!

fathimanourin commented:
I get the following error when i run this 
```
import streamlit as st

file_png = st.file_uploader("Upload a PNG image", type=([".png"]))

if file_png:
    file_png_bytes = st.file_reader(file_png)
    st.image(file_png_bytes)
```

> AttributeError:`` module 'streamlit' has no attribute 'file_reader'
> Traceback:
> File "/home/fathima/.local/lib/python3.8/site-packages/streamlit/script_runner.py", line 338, in _run_script
>     exec(code, module.__dict__)
> File "/home/fathima/Documents/Internship/frontend.py", line 6, in <module>
>     file_png_bytes = st.file_reader(file_png)
